howtos/module/_posts/2016-07-07-Common-mistakes-while-coding-module.md

```diff
@@ -0,0 +1,183 @@
+
+How to restrict your module for specific countries? In your `__construct` method:
+
+`$this->limited_countries = array('fr', 'pl');`
+
+This code will not allow to install the module in countries other than France and Poland. Use [ISO-3166 codes](https://en.wikipedia.org/wiki/ISO_3166-1_alpha-2#Officially_assigned_code_elements)!
+
+
+### 5. Do not trigger code if it is for a specific page
+
+If you are using the `displayHeader` hook to add features to the Product Page (for instance), do not use this code if you're not on page where you have a single product.
+
+You can easily contextualize your code by using simple `if` statements:
+
+{% highlight php %}
+public function hookDisplayHeader($params)
+{
+    // A good practice is to stop code as soon as possible if we don't need it.
+    if ($this->context->controller->php_self !== 'product') {
+        return;
+    }
+
+    // Rest of the code
+}
+{% endhighlight %}
+
+Quick tip: it's worth noting that if you want to get some information from the Product object, you can do this with this code:
+`$product = $this->context->controller->getProduct()`.<br/>
+There is no need to do `new Product(...)`. This code will work on a single-product page.
+
+
+### 6. Too much going on in your __constructor
+
+Yes, this can be tricky. Very often I see modules where developers try to do some stuff in the `__construct` method of their module. It is Not Good. Definitely. Why? Because this code will therefore be triggered on the front office, the back office Modules page, and a few other pages. Be responsible, place your code in hooks, or target a specific page like it was described in point #5.
+
+
+### 7. Conflicts, conflicts everywhere!
+
+So you've created an amazing carousel with products selected from back office, proudly using [owlCarousel](http://www.owlgraphic.com/owlcarousel/), [slick.js](http://kenwheeler.github.io/slick/) or any other popular library. Great!<br/>
+Now a customer bought another carousel module, one with product reviews, and it turns out that this module is using the same library: now nothing is really working.
+
+What can we do to avoid this problem? Well, there's no 100%-safe solution for this problem, but you, as a Good Developer, can be nice to other developers (or theme authors), and allow your customers to disable loading of libraries from your module. Take the time to add a section in your Settings called "Troubleshooting" for example, and create a settings where your user can select which libraries can by turned off from your module.
+
+
+### 8. Your module has front office controllers but you didn't register them...
+
+...and if that's a case, your user cannot set where the left or right column should be displayed. If you're using front office controllers within your module, remember to register them properly in your `__construct` method:
+
+`$this->controllers = array('my-controller', 'other-controller');`
```

Comment:
This is just forgotten soooo often. I believe it's because this is missing in the official documentation, but I'm not sure.